code/game/machinery/status_display.dm

```diff
@@ -148,38 +194,133 @@
 		(current_mode == SD_BLANK) || \
 		(current_mode != SD_PICTURE && message1 == "" && message2 == "") \
 	)
-		set_light(0)
+		set_light(l_on = FALSE)
 		return
-	set_light(1.4, 0.7, LIGHT_COLOR_BLUE) // blue light
+
+	set_light(l_color = text_color, l_on = TRUE)
 
 /obj/machinery/status_display/update_overlays()
 	. = ..()
 
+	// Facing south, we render traditionally.
+	if(dir == SOUTH)
+		add_screen_overlays(.)
+		return
+
+	// Otherwise, we render a projection on the floor.
+
+	// Get screen overlays and return if it's off.
+	var/list/projected_overlays = list()
+	var/anything_displayed = add_screen_overlays(projected_overlays, projection_only = TRUE)
+
+	if(!anything_displayed)
+		return
+
+	// Both of these are vis contents that will receive 1 or more overlays themselves.
+
+	// Matrices to transform the content given the direction
+	var/matrix/floor_matrix = floor_projections["[dir]"]
+	var/matrix/beam_matrix = beam_projections["[dir]"]
+
+	// Make the 2 vis overlays.
+	generate_projection_overlay(floor_matrix, projected_overlays, alpha = PROJECTION_FLOOR_ALPHA)
+	generate_projection_overlay(beam_matrix, projected_overlays, alpha = PROJECTION_BEAM_ALPHA)
+
+	// Impossible for this to look good as an actual emissive since the emissive plane crushes alpha to all or nothing for some reason.
+	// We don't really want anything blocking it anyway since the projection would shine over people.
+	var/mutable_appearance/projection_emissive = mutable_appearance(icon, "projection-mask", offset_spokesman = src, plane = LIGHTING_PLANE)
+	var/list/emissive_offsets = projection_emissive_offsets["[dir]"]
+	projection_emissive.pixel_x = emissive_offsets[1]
+	projection_emissive.pixel_y = emissive_offsets[2]
+	projection_emissive.blend_mode = BLEND_ADD
+	. += projection_emissive
+
+	// Translate the text.
+	message1_overlay?.transform = floor_matrix
+	message1_overlay?.alpha = PROJECTION_TEXT_ALPHA
+	message2_overlay?.transform = floor_matrix
```

Comment:
what is the purpose of these transform sets? doesn't this happen in generate_projection_overlay

Reply:
The `messageX_overlay`s are vis_contents atoms, due to it not being possible to do what the message `maptext` does if they were `mutable_appearance`s. Thus, they are out of band from the projection overlay stuff.

I've added comments and renamed the involved variables so it's more clear.